### ISSUE TITLE
[Blazor] Introduce `[QueryParameter]` and `[FormParameter]`

### ISSUE THREAD

**Issue by javiercn** [dotnet/aspnetcore#48443]
[QueryParameter] replaces `[SupplyFromQueryParameter]` and `[FormParameter]` replaces `[SupplyFromQueryParameter]` (which we just need to rename).

For `[SupplyFromQueryParameter]` we need to keep it for back compatibitily, but we can obsolete it and use our analyzer/code fix to switch to `[QueryParameter]`

**Comment by SteveSandersonMS** [comment]
The new names are great!

> but we can obsolete it

TBH I would also be OK with just leaving it (if it's not actively harming the product). We don't necessarily have to cause the noise of obsoletion warnings if it's just a stylistic change to the naming. The old name happens to be weird enough that once people stop seeing it in examples, they are very unlikely to discover and start using it.

**Comment by davidfowl** [comment]
If we are going to do this I wonder if we can reuse these attributes for minimal APIs? 

cc @captainsafia @DamianEdwards

**Comment by DamianEdwards** [comment]
Hmmm interesting. Here's how it would look I assume:

```csharp
app.MapGet("/search", async ([QueryParameter] string search, [QueryParameter] int? page, [FromServices] ISearchService searchService) =>
{
    return await searchService.GetResults(search, page);
});

app.MapPost("/cart", async ([FormParameter] id productId, [FormParameter] int qty, [FromServices] ICartService cartService) =>
{
    return await cartService.AddItem(productId, qty);
});
```

Or using `AsParameters` with a record:

```csharp
app.MapGet("/search", async ([AsParameters] SearchInput input, [FromServices] ISearchService searchService) =>
{
    return await searchService.GetResults(input);
});

record struct SearchInput([QueryParameter] string SearchTerm, [QueryParameter] int Page = 1);
```

**Comment by javiercn** [comment]
> If we are going to do this I wonder if we can reuse these attributes for minimal APIs?

I don't think we can, these attributes need to live in `Microsoft.AspNetCore.Components` and `Microsoft.AspNetCore.Components.Web` and I would prefer we avoid doing something that precludes us from adding component specific semantics to them in the future.

**Comment by javiercn** [comment]
> Hmmm interesting. Here's how it would look I assume:

> app.MapGet("/search", async ([QueryParameter] string search, [QueryParameter] int? page, [FromServices] ISearchService searchService) =>
{
    return await searchService.GetResults(search, page);
});

> app.MapPost("/cart", async ([FormParameter] id productId, [FormParameter] int qty, [FromServices] ICartService cartService) =>
{
    return await cartService.AddItem(productId, qty);
});
Or using AsParameters with a record:

> app.MapGet("/search", async ([AsParameters] SearchInput input, [FromServices] ISearchService searchService) =>
{
    return await searchService.GetResults(input);
});

> record struct SearchInput([QueryParameter] string SearchTerm, [QueryParameter] int Page = 1);

I don't think we envision these being used outside of the Blazor context. The renderer needs to be aware of them, and we want to have the choice in the future to add Blazor specific idioms to them.

**Comment by davidfowl** [comment]
I don’t think so. We have this metadata assembly today that we us me to share attributes between blazor and the rest of the stack. These could love there.

**Comment by javiercn** [comment]
@davidfowl will we be ok if we need to put Blazor specific information in parameters in the future?

**Comment by davidfowl** [comment]
Blazor has been around for 5 years now, what sorts of blazor specific parameters do you envision ?

**Comment by captainsafia** [comment]
> If we are going to do this I wonder if we can reuse these attributes for minimal APIs?
> 
> cc @captainsafia @DamianEdwards

@javiercn shared my initial reaction to this statement which is that we would have to put these attributes lower in the stack in order to share them between Blazor and minimal APIs.

I have a concern about the bifurcation we'll introduce if two parameter sources (query and form) depend on a set of attributes that are lower in the stack compared to the other attributes that minimal APIs relies on. 

The code samples that @DamianEdwards shared heighten my reservations here. It looks _really_ weird to have attributes with two different styles in the route handler. `([FormParameter] IFormFile fromForm, [QueryParameter] int page, [FromServices] TodoService service, [FromHeader] int timeout)`.

TL;DR: I'm not keen on it...

**Comment by davidfowl** [comment]
@captainsafia we have razor pages, mvc, minimal APIs and now blazor SSR. They all have similar features yet we have these different attributes because we don't have a time machine. Now we're going to introduce new unique attributes into Blazor because we need them and we don't have time to think through the implications of unifying them.

We had this exact same discussion when we were doing minimal APIs with the MVC attributes and ended up reusing them. If we're going to make new attributes, let' see if we can do a better job now that we have the benefit of hindsight.

**Comment by javiercn** [comment]
@davidfowl it's a bit more nuanced than that.

I'm happy if we discuss this more when @SteveSandersonMS is back. It's a bit more nuanced than that because we try to avoid bundling web related concepts into Microsoft.AspNetCore.Components in case we use it on a separate context (like mobile bindings).

So, if we need to have some "shared" attribute, we'll have to think about how we do that in a way that doesn't bring web related stuff into M.A.C.

For example, we would need to make M.A.C.Web depend on this M.A.M instead, but at the same time, some closed parts (like cascading values) need to be able to consume things from M.A.C.Web in a way that is not "open" to everyone for extension.

Right now, this works because M.A.C.Web has IVT to M.A.C and we can define internal interfaces there so that we are the only ones that can consume or create types implementing them.

Not saying that it's not possible, just pointing out that it might require a lot of rework on the components side, and I'm worried of the cost it will have compared to the benefit.

Are there other alternatives that we could explore? What if we were to use type aliases on the file for this? Ultimately, is very unlikely that you have a Razor component and a minimal API defined on the same file. Similarly with controllers in MVC.

**Comment by captainsafia** [comment]
> Now we're going to introduce new unique attributes into Blazor because we need them and we don't have time to think through the implications of unifying them.

Well, yes, although to be frank sharing these specific attributes doesn't address the unification problem altogether. There's still other attributes that would need this unification treatment that aren't `FromQuery` or `FromForm`. I recognize the problem you are calling out but I don't think this is the appropriate avenue for solving it.

I do think it might be worthwhile to design `QueryParameter` and `FormParameter` with an eye as to how they can be used throughout the stack, but that doesn't solve our attribute problem all up.

> It's a bit more nuanced than that because we try to avoid bundling web related concepts into Microsoft.AspNetCore.Components in case we use it on a separate context (like mobile bindings).

This constraint makes this a little more challenging because we're talking about platform-agnostic attributes. There are some attributes that specifically reference HTTP concepts (`FromHeader`).

I do wonder what `QueryParameter` means in the context of mobile bindings though?

Also, from the original issue description:

> [QueryParameter] replaces [SupplyFromQueryParameter] and [FormParameter] replaces [SupplyFromQueryParameter] (which we just need to rename).

Is there a typo here? Should the second `SupplyFromQueryParameter` be `SupplyFromFormParameter`?

One other thing to call out here is that if we _do_ unify the attributes, we should probably have a plan for when deprecating the existing onces.

**Comment by SteveSandersonMS** [comment]
> I do wonder what QueryParameter means in the context of mobile bindings though?

For native apps, there's a notion of deep linking that uses URLs (with a custom scheme) to send the user to a particular part of the app. And for Android specifically, URLs are used more deeply to track a user's position in a navigation stack so the "back" button works. So query parameters are relevant for native apps. But as far as I know, form parameters or HTTP headers don't have any relevant meaning within native apps, since there's no HTTP server.

> Is there a typo here? Should the second SupplyFromQueryParameter be SupplyFromFormParameter?

Yes, that's a typo :)

More generally, I certainly see the various ways this could be tricky. I suspect we *can* find a way to share attributes if we want, especially with the proposals being based on being entirely new attributes. That means they probably don't have to go into M.A.Components and we wouldn't necessarily have such difficult layering issues on the Blazor side (I'm hoping they could just go in .Web or perhaps some new abstractions assembly that is referenced from the app SDK by default and can be referenced optionally by RCLs).

Has there been consideration of reusing the existing `[FromQuery]` and `[FromForm]`, pushing them down to a lower assembly via type forwarding, so we don't even need something new? I won't be surprised if people say that's impractical but asking at least to fill out the discussion.
